pkgs/racket-test/tests/racket/contract/contract-out.rkt

```diff
@@ -1958,4 +1958,33 @@
                #`'#,(struct-field-info-list (syntax-local-value #'foo))))
       (eval '(extract-field-names)))
    (list 'x))
+
+  (test/spec-passed/result
+   'provide/contract-upe-struct-transformer
+   '(begin
+      (eval '(module provide/contract-upe-struct-transformer racket/base
+               (require racket/contract/base)
+               (provide (contract-out
+                         #:unprotected-submodule no-contract
+                         (struct s ([x integer?]))))
+               (struct s (x) #:constructor-name make-s)))
+      (eval '(require racket/match
+                      (submod 'provide/contract-upe-struct-transformer no-contract)))
+      (eval '(match (make-s 1) [(s x) x])))
+   1)
+
+  (test/spec-passed/result
+   'provide/contract-upe-struct-type-descriptor
+   '(begin
+      (eval '(module provide/contract-upe-struct-type-descriptor racket/base
+               (require racket/contract/base)
+               (provide (contract-out
+                         #:unprotected-submodule no-contract
+                         (struct s ([x integer?]))))
+               (struct s (x))))
+      (eval '(require racket/match
+                      (submod 'provide/contract-upe-struct-type-descriptor no-contract)))
+      (eval '(struct t s (y)))
+      (eval '(match (t 1 2) [(t x y) (list x y)])))
+   (list 1 2))
```

Comment:
Ditto here. To test a result, you can have the module export a definition with a name that has a prefix of `provide/contract-upe-struct-type-descriptor`.